pkgs/top-level/python-packages.nix

```diff
@@ -15329,6 +15329,35 @@ self: super: with self; {
       protobufTF = pkgs.protobuf_21.override {
         abseil-cpp = pkgs.abseil-cpp_202301;
       };
+      ml-dtypesTF = (self.ml-dtypes.overrideAttrs (
+        oldAttrs: rec {
+          # recent versions deprecated float8_e4m3b11
+          version = "0.1.0";
+          name = "${oldAttrs.pname}-${version}";
+          src = pkgs.fetchFromGitHub {
+            owner = "jax-ml";
+            repo = "ml_dtypes";
+            rev = "v${version}";
+            hash = "sha256-3rs48WtXAfP5g15j8BDSd0ee+c6CPy+OTfUB3HLgPm8=";
+            fetchSubmodules = true;
+          };
+          buildInputs = [ pybind11 ];
+          dependencies = [ numpy ];
+          postPatch = ''
+            substituteInPlace pyproject.toml \
+              --replace-fail "numpy~=1.21.2" "numpy" \
+              --replace-fail "numpy~=1.23.3" "numpy" \
+              --replace-fail "pybind11~=2.10.0" "pybind11" \
+              --replace-fail "setuptools~=67.6.0" "setuptools"
+          '';
+          disabledTests = oldAttrs.disabledTests or []
+          # these fail on Darwin:
+          # https://github.com/jax-ml/ml_dtypes/issues/47#issuecomment-1483821629
+          ++ lib.optionals stdenv.isDarwin [
+            "testBetweenCustomTypes_bfloat16" "testPredicateUfunc_bfloat16"
+          ];
+        })
+      );
```

Comment:
@zeuner would it do you a terrible pain with the subsequent PRs if I moved this whole let-in into formal parameters in `pkgs/development/python-modules/tensorflow/default.nix`? I'd like to just `callPackage` the modified path in an overlay for faster rebuilds, but also in general I don't think we should be offloading this kind of complexity from the package "recipe" onto caller

Reply:
Sounds reasonable. Would you have a commit available for what you have in mind?

Reply:
@SomeoneSerge Any news regarding this?

Frankly, I'm not fully confident about this anymore. Maybe at least parts of the packages pinned here should be considered to be part of `tensorflow`'s ABI, considering that packages that interact with it on a low level expect compatible versions. This can be seen with `keras`, which would only build when using the same `ml-dtypes` as `tensorflow` (https://github.com/NixOS/nixpkgs/pull/272838/commits/5cafb999dc930ee64ac57ca9456e8125ea5943f6 - I'm currently tackling this by exporting the compatible version through `passthru`), and I think it's also plausible (not tried yet) that `tf2onnx` might not have to be marked as broken (https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/tf2onnx/default.nix#L80) if the maintainer tried to use the same pinning an `tensorflow` itself. An alternative to hiding away such interface-relevant details might be to make it easier to pick up the pinnings for other packages.

That being said, it seems that most pinnings become unnecessary starting from `2.16`.

Reply:
Update: As of the latest `master` branch, `2.16` still requires the `protobuf` pinning. Currently investigating this. I still assume the expression will be smaller on subsequent updates.